runtime/src/accounts_index.rs

```diff
@@ -1,38 +1,297 @@
+    fn iter<R>(&self, range: Option<R>) -> AccountsIndexIterator<T>
+    where
+        R: RangeBounds<Pubkey>,
+    {
+        AccountsIndexIterator::new(&self.account_maps, range)
+    }
+
+    fn do_scan_accounts<'a, F, R>(&'a self, ancestors: &Ancestors, mut func: F, range: Option<R>)
     where
         F: FnMut(&Pubkey, (&T, Slot)),
-        I: Iterator<Item = (&'a Pubkey, &'a AccountMapEntry<T>)>,
+        R: RangeBounds<Pubkey>,
     {
-        for (pubkey, list) in iter {
-            let list_r = &list.1.read().unwrap();
-            if let Some(index) = self.latest_slot(Some(ancestors), &list_r, None) {
-                func(pubkey, (&list_r[index].1, list_r[index].0));
+        for pubkey_list in self.iter(range) {
+            for (pubkey, list) in pubkey_list {
+                let list_r = &list.slot_list.read().unwrap();
+                if let Some(index) = self.latest_slot(Some(ancestors), &list_r, None) {
+                    func(&pubkey, (&list_r[index].1, list_r[index].0));
+                }
+            }
+        }
+    }
+
+    pub fn get_account_read_entry(&self, pubkey: &Pubkey) -> Option<ReadAccountMapEntry<T>> {
+        self.account_maps
+            .read()
+            .unwrap()
+            .get(pubkey)
+            .cloned()
+            .map(ReadAccountMapEntry::from_account_map_entry)
+    }
+
+    fn get_account_write_entry(&self, pubkey: &Pubkey) -> Option<WriteAccountMapEntry<T>> {
+        self.account_maps
+            .read()
+            .unwrap()
+            .get(pubkey)
+            .cloned()
+            .map(WriteAccountMapEntry::from_account_map_entry)
+    }
+
+    fn get_account_write_entry_else_create(
+        &self,
+        pubkey: &Pubkey,
+    ) -> (WriteAccountMapEntry<T>, bool) {
+        let mut w_account_entry = self.get_account_write_entry(pubkey);
+        let mut is_newly_inserted = false;
+        if w_account_entry.is_none() {
+            let new_entry = Arc::new(AccountMapEntryInner {
+                ref_count: AtomicU64::new(0),
+                slot_list: RwLock::new(SlotList::with_capacity(32)),
+            });
+            let mut w_account_maps = self.account_maps.write().unwrap();
+            let account_entry = w_account_maps.entry(*pubkey).or_insert_with(|| {
+                is_newly_inserted = true;
+                new_entry
+            });
+            w_account_entry = Some(WriteAccountMapEntry::from_account_map_entry(
+                account_entry.clone(),
+            ));
+        }
+
+        (w_account_entry.unwrap(), is_newly_inserted)
+    }
+
+    pub fn handle_dead_keys(&self, dead_keys: Vec<Pubkey>) {
+        if !dead_keys.is_empty() {
+            for key in &dead_keys {
+                let mut w_index = self.account_maps.write().unwrap();
+                if let Some(account_entry) = w_index.get(key) {
+                    if account_entry.slot_list.read().unwrap().is_empty() {
+                        w_index.remove(key);
+                    }
```

Comment:
nits: I think we could reduce the lookup from 2 times to once like this:

```rust
if let Some(index_entry) = w_index.entry(key) {
    if (...) {
        index_entry.remove();
    }
}
```

Reply:
slight variation on this using `Occupied` instead of `Some`, but updated!